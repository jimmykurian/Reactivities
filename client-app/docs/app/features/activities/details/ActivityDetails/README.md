[**Client App Documentation v0.0.0**](../../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../../README.md) / app/features/activities/details/ActivityDetails

# app/features/activities/details/ActivityDetails

## Index

### Interfaces

- [Props](interfaces/Props.md)

### Functions

- [default](functions/default.md)
