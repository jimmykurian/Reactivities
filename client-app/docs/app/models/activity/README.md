[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/models/activity

# app/models/activity

## Index

### Interfaces

- [Activity](interfaces/Activity.md)
