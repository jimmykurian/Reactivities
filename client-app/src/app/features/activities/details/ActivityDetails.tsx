/**
 * @file ActivityDetails.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityDetails component.
 */

import { Grid } from 'semantic-ui-react';
import { useStore } from '../../../stores/store';
import LoadingComponent from '../../../layout/LoadingComponent';
import { observer } from 'mobx-react-lite';
import { useParams } from 'react-router-dom';
import { useEffect } from 'react';
import ActivityDetailedChat from './ActivityDetailedChat';
import ActivityDetailedHeader from './ActivityDetailedHeader';
import ActivityDetailedInfo from './ActivityDetailedInfo';
import ActivityDetailedSidebar from './ActivityDetailedSidebar';

/**
 * @component ActivityDetails
 * @description A functional component that displays the details of an activity.
 *
 * @returns {JSX.Element} The JSX element representing the activity details.
 *
 * @remarks
 * This component uses the Semantic UI React library to create a layout for displaying detailed information about an activity.
 * It includes a header, detailed information, a chat section, and a sidebar.
 * The `loadActivity` function is called to fetch the activity details based on the ID from the URL parameters.
 * The component accesses the `activityStore` from the MobX store context to get the selected activity and relevant functions.
 * The `LoadingComponent` is displayed while the activity details are being loaded.
 *
 * @example
 * Here is an example of how to use the ActivityDetails component:
 * ```tsx
 * import React from 'react';
 * import ActivityDetails from './features/activities/details/ActivityDetails';
 * import { BrowserRouter as Router, Route } from 'react-router-dom';
 * import { StoreProvider } from './stores/store';
 *
 * const App = () => (
 *   <StoreProvider>
 *     <Router>
 *       <Route path="/activities/:id" component={ActivityDetails} />
 *     </Router>
 *   </StoreProvider>
 * );
 *
 * export default App;
 * ```
 */
export default observer(function ActivityDetails(): JSX.Element {
  const { activityStore } = useStore();
  const {
    selectedActivity: activity,
    loadActivity,
    loadingInitial,
  } = activityStore;
  const { id } = useParams();

  useEffect(() => {
    if (id) loadActivity(id);
  }, [id, loadActivity]);

  if (loadingInitial || !activity)
    return <LoadingComponent content="Loading app..." />;

  return (
    <Grid>
      <Grid.Column width={10}>
        <ActivityDetailedHeader activity={activity} />
        <ActivityDetailedInfo activity={activity} />
        <ActivityDetailedChat />
      </Grid.Column>
      <Grid.Column width={6}>
        <ActivityDetailedSidebar />
      </Grid.Column>
    </Grid>
  );
});
