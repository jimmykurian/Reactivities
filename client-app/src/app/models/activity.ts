/**
 * @author Jimmy Kurian
 * @name Activity
 */

/**
 * The Activity interface represents the structure of an activity object in the Reactivities application.
 * It includes various properties such as id, title, date, description, category, city, and venue.
 *
 * @interface
 *
 * @property {string} id - The unique identifier for the activity.
 * @property {string} title - The title of the activity.
 * @property {string} date - The date of the activity in ISO string format.
 * @property {string} description - A brief description of the activity.
 * @property {string} category - The category to which the activity belongs.
 * @property {string} city - The city where the activity takes place.
 * @property {string} venue - The specific venue of the activity within the city.
 */
export interface Activity {
  id: string;
  title: string;
  date: string;
  description: string;
  category: string;
  city: string;
  venue: string;
}
