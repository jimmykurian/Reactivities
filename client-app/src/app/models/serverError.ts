/**
 * @file ServerError.ts
 * @author Jimmy Kurian
 * @fileoverview This file defines the `ServerError` interface, which models the structure of server error responses in the Reactivities application.
 */

/**
 * The ServerError interface represents the structure of an error returned from the server.
 * It includes the HTTP status code, an error message, and optional detailed information about the error.
 *
 * @interface ServerError
 *
 * @property {number} statusCode - The HTTP status code associated with the error.
 * @property {string} message - A brief message describing the error.
 * @property {string} details - Additional details or stack trace information about the error (optional).
 */
export interface ServerError {
  statusCode: number;
  message: string;
  details: string;
}
