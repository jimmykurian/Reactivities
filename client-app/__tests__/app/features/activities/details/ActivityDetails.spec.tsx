import { render, screen } from '@testing-library/react';
import { faker } from '@faker-js/faker';
import { Activity } from '../../../../../src/app/models/activity';
import ActivityDetails from '../../../../../src/app/features/activities/details/ActivityDetails';

describe('ActivityDetails', () => {
  // Generate a mock activity using Faker.js
  const generateMockActivity = (): Activity => ({
    id: faker.string.uuid(),
    title: faker.lorem.words(3),
    date: faker.date.future().toISOString(),
    description: faker.lorem.sentence(),
    category: faker.lorem.word(),
    city: faker.location.city(),
    venue: faker.location.streetAddress(),
  });

  // Static mock activity for consistent snapshot tests
  const staticMockActivity: Activity = {
    id: '1',
    title: 'Static Title',
    date: '2024-06-12T00:00:00.000Z',
    description: 'This is a static description for testing.',
    category: 'static-category',
    city: 'Static City',
    venue: 'Static Venue',
  };

  const cancelSelectActivity = jest.fn();
  const openForm = jest.fn();

  test('renders the ActivityDetails component', () => {
    // Arrange
    const mockActivity = generateMockActivity();
    render(
      <ActivityDetails
        activity={mockActivity}
        cancelSelectActivity={cancelSelectActivity}
        openForm={openForm}
      />,
    );

    // Act & Assert
    expect(screen.getByText(mockActivity.title)).toBeInTheDocument();
    expect(screen.getByText(mockActivity.date)).toBeInTheDocument();
    expect(screen.getByText(mockActivity.description)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Edit/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Cancel/i })).toBeInTheDocument();
  });

  test('displays the correct image for the activity category', () => {
    // Arrange
    const mockActivity = generateMockActivity();
    render(
      <ActivityDetails
        activity={mockActivity}
        cancelSelectActivity={cancelSelectActivity}
        openForm={openForm}
      />,
    );

    // Act & Assert
    const imgElement = screen.getByRole('img');
    expect(imgElement).toHaveAttribute(
      'src',
      `/assets/categoryImages/${mockActivity.category}.jpg`,
    );
  });

  test('matches snapshot', () => {
    // Arrange
    const { asFragment } = render(
      <ActivityDetails
        activity={staticMockActivity}
        cancelSelectActivity={cancelSelectActivity}
        openForm={openForm}
      />,
    );

    // Act & Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
