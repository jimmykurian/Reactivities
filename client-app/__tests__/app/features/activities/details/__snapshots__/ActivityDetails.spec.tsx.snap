// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityDetails matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="ten wide column"
    >
      <div
        class="ui segments"
      >
        <div
          class="ui basic top attached segment"
          style="padding: 0px;"
        >
          <img
            class="ui fluid image"
            src="/assets/categoryImages/static-category.jpg"
            style="filter: brightness(30%);"
          />
          <div
            class="ui basic segment"
            style="position: absolute; bottom: 5%; left: 5%; width: 100%; height: auto; color: white;"
          >
            <div
              class="ui items"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="ui huge header"
                    style="color: white;"
                  >
                    Static Title
                  </div>
                  <p>
                    2024-06-12T00:00:00.000Z
                  </p>
                  <p>
                    Hosted by 
                    <strong>
                      Bob
                    </strong>
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ui clearing bottom attached segment"
        >
          <button
            class="ui teal button"
          >
            Join Activity
          </button>
          <button
            class="ui button"
          >
            Cancel attendance
          </button>
          <button
            class="ui orange right floated button"
          >
            Manage Event
          </button>
        </div>
      </div>
      <div
        class="ui segments"
      >
        <div
          class="ui top attached segment"
        >
          <div
            class="ui grid"
          >
            <div
              class="one wide column"
            >
              <i
                aria-hidden="true"
                class="teal info large icon"
              />
            </div>
            <div
              class="fifteen wide column"
            >
              <p>
                This is a static description for testing.
              </p>
            </div>
          </div>
        </div>
        <div
          class="ui attached segment"
        >
          <div
            class="ui middle aligned grid"
          >
            <div
              class="one wide column"
            >
              <i
                aria-hidden="true"
                class="teal calendar large icon"
              />
            </div>
            <div
              class="fifteen wide column"
            >
              <span>
                2024-06-12T00:00:00.000Z
              </span>
            </div>
          </div>
        </div>
        <div
          class="ui attached segment"
        >
          <div
            class="ui middle aligned grid"
          >
            <div
              class="one wide column"
            >
              <i
                aria-hidden="true"
                class="teal marker large icon"
              />
            </div>
            <div
              class="eleven wide column"
            >
              <span>
                Static Venue, Static City
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ui teal inverted top attached center aligned segment"
      >
        <div
          class="ui header"
        >
          Chat about this event
        </div>
      </div>
      <div
        class="ui attached segment"
      >
        <div
          class="ui comments"
        >
          <div
            class="comment"
          >
            <div
              class="avatar"
            >
              <img
                src="/assets/user.png"
              />
            </div>
            <div
              class="content"
            >
              <a
                class="author"
              >
                Matt
              </a>
              <div
                class="metadata"
              >
                <div>
                  Today at 5:42PM
                </div>
              </div>
              <div
                class="text"
              >
                How artistic!
              </div>
              <div
                class="actions"
              >
                <a
                  class=""
                >
                  Reply
                </a>
              </div>
            </div>
          </div>
          <div
            class="comment"
          >
            <div
              class="avatar"
            >
              <img
                src="/assets/user.png"
              />
            </div>
            <div
              class="content"
            >
              <a
                class="author"
              >
                Joe Henderson
              </a>
              <div
                class="metadata"
              >
                <div>
                  5 days ago
                </div>
              </div>
              <div
                class="text"
              >
                Dude, this is awesome. Thanks so much
              </div>
              <div
                class="actions"
              >
                <a
                  class=""
                >
                  Reply
                </a>
              </div>
            </div>
          </div>
          <form
            class="ui reply form"
          >
            <div
              class="field"
            >
              <textarea
                rows="3"
              />
            </div>
            <button
              class="ui icon primary left labeled button"
            >
              <i
                aria-hidden="true"
                class="edit icon"
              />
              Add Reply
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="six wide column"
    >
      <div
        class="ui teal inverted secondary top attached center aligned segment"
      >
        3 People Going
      </div>
      <div
        class="ui attached segment"
      >
        <div
          class="ui divided relaxed list"
          role="list"
        >
          <div
            class="item"
            style="position: relative;"
          >
            <div
              class="ui orange right ribbon label"
              style="position: absolute;"
            >
              Host
            </div>
            <img
              class="ui tiny image"
              src="/assets/user.png"
            />
            <div
              class="middle aligned content"
            >
              <h3
                class="header"
              >
                <a
                  href="/activities/1"
                >
                  Bob
                </a>
              </h3>
              <div
                class="extra"
                style="color: orange;"
              >
                Following
              </div>
            </div>
          </div>
          <div
            class="item"
            style="position: relative;"
          >
            <img
              class="ui tiny image"
              src="/assets/user.png"
            />
            <div
              class="middle aligned content"
            >
              <h3
                class="header"
              >
                <a
                  href="/activities/1"
                >
                  Tom
                </a>
              </h3>
              <div
                class="extra"
                style="color: orange;"
              >
                Following
              </div>
            </div>
          </div>
          <div
            class="item"
            style="position: relative;"
          >
            <img
              class="ui tiny image"
              src="/assets/user.png"
            />
            <div
              class="middle aligned content"
            >
              <h3
                class="header"
              >
                <a
                  href="/activities/1"
                >
                  Sally
                </a>
              </h3>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
